src/lib/drivers/magnetometer/PX4Magnetometer.cpp

```diff
@@ -34,118 +34,100 @@
 
 #include "PX4Magnetometer.hpp"
 
-PX4Magnetometer::PX4Magnetometer(const char *path, device::Device  *interface, uint8_t dev_type, enum Rotation rotation,
-				 float scale) :
-	CDev(path),
-	_interface(interface)
-{
-	_device_id.devid = _interface->get_device_id();
-	// _device_id.devid_s.bus_type = (device::Device::DeviceBusType)_interface->get_device_bus_type();
-	// _device_id.devid_s.bus = _interface->get_device_bus();
-	// _device_id.devid_s.address = _interface->get_device_address();
-	_device_id.devid_s.devtype = dev_type;
-
-	CDev::init();
+#include <lib/drivers/device/Device.hpp>
 
+PX4Magnetometer::PX4Magnetometer(uint32_t device_id, uint8_t priority, enum Rotation rotation) :
+	CDev(nullptr),
+	_sensor_mag_pub{ORB_ID(sensor_mag), priority},
+	_rotation{rotation}
+{
 	_class_device_instance = register_class_devname(MAG_BASE_DEVICE_PATH);
 
-	_rotation = rotation;
-	_scale = scale;
+	_sensor_mag_pub.get().device_id = device_id;
+	_sensor_mag_pub.get().scaling = 1.0f;
 
-	_cal.x_offset = 0;
-	_cal.x_scale  = 1.0f;
-	_cal.y_offset = 0;
-	_cal.y_scale  = 1.0f;
-	_cal.z_offset = 0;
-	_cal.z_scale  = 1.0f;
+	// force initial publish to allocate uORB buffer
+	// TODO: can be removed once all drivers are in threads
+	_sensor_mag_pub.update();
 }
 
 PX4Magnetometer::~PX4Magnetometer()
 {
-	if (_topic != nullptr) {
-		orb_unadvertise(_topic);
-	}
-}
-
-int PX4Magnetometer::init()
-{
-	mag_report report{};
-	report.device_id = _device_id.devid;
-
-	if (_topic == nullptr) {
-		_topic = orb_advertise_multi(ORB_ID(sensor_mag), &report, &_orb_class_instance, ORB_PRIO_HIGH - 1);
-
-		if (_topic == nullptr) {
-			PX4_ERR("Advertise failed.");
-			return PX4_ERROR;
-		}
+	if (_class_device_instance != -1) {
+		unregister_class_devname(MAG_BASE_DEVICE_PATH, _class_device_instance);
 	}
-
-	return PX4_OK;
 }
 
 int PX4Magnetometer::ioctl(cdev::file_t *filp, int cmd, unsigned long arg)
 {
 	switch (cmd) {
-	case MAGIOCSSCALE:
-		// Copy scale in.
-		memcpy(&_cal, (struct mag_calibration_s *) arg, sizeof(_cal));
-		return OK;
+	case MAGIOCSSCALE: {
+			// Copy offsets and scale factors in
+			mag_calibration_s cal{};
+			memcpy(&cal, (mag_calibration_s *) arg, sizeof(cal));
+
+			_calibration_offset = matrix::Vector3f{cal.x_offset, cal.y_offset, cal.z_offset};
+			_calibration_scale = matrix::Vector3f{cal.x_scale, cal.y_scale, cal.z_scale};
+		}
 
-	case MAGIOCGSCALE:
-		// Copy scale out.
-		memcpy((struct mag_calibration_s *) arg, &_cal, sizeof(_cal));
-		return OK;
+		return PX4_OK;
 
 	case DEVIOCGDEVICEID:
-		return _device_id.devid;
+		return _sensor_mag_pub.get().device_id;
```

Comment:
not device_id.devid? 
is this in support of multi instance?

Reply:
This is the raw int and the only thing stored in this class.